velox/exec/tests/MultiFragmentTest.cpp

```diff
@@ -1448,5 +1448,157 @@ DEBUG_ONLY_TEST_F(MultiFragmentTest, mergeWithEarlyTermination) {
+
+  std::shared_ptr<PartitionedOutputBufferManager> bufferManager() {
+    return PartitionedOutputBufferManager::getInstance().lock();
+  }
+
+  void doFetch(int64_t sequence) {
+    bool ok = bufferManager()->getData(
+        taskId_,
+        destination_,
+        maxBytes_,
+        sequence,
+        [&](auto pages, auto sequence) mutable {
+          const auto nextSequence = sequence + pages.size();
+          const bool atEnd = processData(std::move(pages), sequence);
+          bufferManager()->acknowledge(taskId_, destination_, nextSequence);
+
+          if (atEnd) {
+            bufferManager()->deleteResults(taskId_, destination_);
+            promise_.setValue();
+          } else {
+            doFetch(nextSequence);
+          }
+        });
+    VELOX_CHECK(ok);
+  }
+
+  bool processData(
+      std::vector<std::unique_ptr<folly::IOBuf>> pages,
+      int64_t sequence) {
+    numPages_ += pages.size();
+    ++numPackets_;
+
+    int64_t numBytes = 0;
+    bool atEnd = false;
+    for (const auto& page : pages) {
+      if (page == nullptr) {
+        VELOX_CHECK(!atEnd);
+        atEnd = true;
+      } else {
+        numBytes += page->computeChainDataLength();
+      }
+    }
+
+    totalBytes_ += numBytes;
+    return atEnd;
+  }
+
+  const std::string taskId_;
+  const int32_t destination_;
+  const int64_t maxBytes_;
+  ContinuePromise promise_{ContinuePromise::makeEmpty()};
+  int32_t numPackets_{0};
+  int32_t numPages_{0};
+  int64_t totalBytes_{0};
+};
+
+/// Verify that POBM::getData() honors maxBytes parameter roughly at 1MB
+/// granularity. It can do so only if PartitionedOutput operator limits the size
+/// of individual pages. PartitionedOutput operator is expected to limit page
+/// sizes to no more than 1MB give and take 30%.
+TEST_F(MultiFragmentTest, maxBytes) {
+  std::string s(25, 'x');
+  auto data = makeRowVector({
+      makeFlatVector<int64_t>(10'000, [](auto row) { return row; }),
+      makeConstant(StringView(s), 10'000),
+  });
+
+  auto plan = PlanBuilder()
+                  .values({data}, false, 100)
+                  .partitionedOutput({}, 1)
+                  .planNode();
+
+  int32_t testIteration = 0;
+  DataFetcher::Stats prevStats;
+
+  auto test = [&](int64_t maxBytes) {
+    const auto taskId = fmt::format("test.{}", testIteration++);
+
+    SCOPED_TRACE(taskId);
+    auto task = makeTask(taskId, plan, 0);
+    Task::start(task, 1);
+    task->updateOutputBuffers(1, true);
+
+    // Allow for data to accumulate.
+    std::this_thread::sleep_for(std::chrono::milliseconds(500));
+
+    DataFetcher fetcher(taskId, 0, maxBytes);
+    fetcher.fetch().wait();
+
+    ASSERT_TRUE(waitForTaskCompletion(task.get()));
+
+    auto stats = fetcher.stats();
+    if (testIteration > 1) {
+      ASSERT_EQ(prevStats.numPages, stats.numPages);
+      ASSERT_EQ(prevStats.totalBytes, stats.totalBytes);
+      ASSERT_GT(prevStats.numPackets, stats.numPackets);
```

Comment:
Why the number of packets gets reduced? Thanks!

Reply:
I'm increasing maxBytes, so expect larger packets.